xds/internal/xdsclient/tests/ads_stream_watch_test.go

```diff
@@ -0,0 +1,197 @@
+	ldsCancel2 := xdsresource.WatchListener(client, listenerName2, noopListenerWatcher{})
+	defer ldsCancel2()
+	if err := waitForResourceWatchState(ctx, client, listenerName2, ads.ResourceWatchStateRequested, true); err != nil {
+		t.Fatal(err)
+	}
+
+	// Stop the server to break the ADS stream. Since the first resource was
+	// already received, this should not change anything for it. But for the
+	// second resource, it should result in the timer getting stopped and the
+	// watch state moving to `started`.
+	lis.Stop()
+	if err := waitForResourceWatchState(ctx, client, listenerName2, ads.ResourceWatchStateStarted, false); err != nil {
+		t.Fatal(err)
+	}
+	if err := verifyResourceWatchState(client, listenerName1, ads.ResourceWatchStateReceived, false); err != nil {
+		t.Fatal(err)
+	}
+
+	// Restart the server and verify that the timer is running and the watch
+	// state is `requested`.
+	lis.Restart()
+	if err := waitForResourceWatchState(ctx, client, listenerName2, ads.ResourceWatchStateRequested, true); err != nil {
+		t.Fatal(err)
+	}
+
+	// Configure the second resource on the management server. This should result
+	// in the resource being pushed to the xDS client and should result in the
+	// timer getting stopped and the watch state moving to `received`.
+	listenerResource2 := e2e.DefaultClientListener(listenerName2, routeConfigName)
+	resources = e2e.UpdateOptions{
+		NodeID:         nodeID,
+		Listeners:      []*v3listenerpb.Listener{listenerResource1, listenerResource2},
+		SkipValidation: true,
+	}
+	if err := mgmtServer.Update(ctx, resources); err != nil {
+		t.Fatal(err)
+	}
+	if err := waitForResourceWatchState(ctx, client, listenerName2, ads.ResourceWatchStateReceived, false); err != nil {
+		t.Fatal(err)
+	}
+}
+
+func (s) TestADS_WatchState_TimerFires(t *testing.T) {
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+
+	// Start an xDS management server.
+	mgmtServer := e2e.StartManagementServer(t, e2e.ManagementServerOptions{})
+
+	// Create an xDS client with bootstrap pointing to the above server, and a
+	// short resource expiry timeout.
+	nodeID := uuid.New().String()
+	bc := e2e.DefaultBootstrapContents(t, nodeID, mgmtServer.Address)
+	testutils.CreateBootstrapFileForTesting(t, bc)
+	client, close, err := xdsclient.NewForTesting(xdsclient.OptionsForTesting{
+		Name:               t.Name(),
+		Contents:           bc,
+		WatchExpiryTimeout: defaultTestWatchExpiryTimeout,
+	})
+	if err != nil {
+		t.Fatalf("Failed to create xDS client: %v", err)
+	}
+	defer close()
+
+	// Create a watch for the first listener resource and verify that the timer
+	// is running and the watch state is `requested`.
+	const listenerName = "listener"
+	ldsCancel1 := xdsresource.WatchListener(client, listenerName, noopListenerWatcher{})
+	defer ldsCancel1()
+	if err := waitForResourceWatchState(ctx, client, listenerName, ads.ResourceWatchStateRequested, true); err != nil {
+		t.Fatal(err)
+	}
+
```

Comment:
Mention watch expiry hits so expected to go timeout.

Reply:
Done.